File: AstroSleep-iOS/AstroSleep/Core/Models/ElementVector.swift
import Foundation

// MARK: - Element Vector
/// [Fire, Earth, Air, Water] scoring vector with decimal precision
struct ElementVector: Codable, Equatable, AdditiveArithmetic {
    var fire: Double
    var earth: Double
    var air: Double
    var water: Double
    
    static let zero = ElementVector(fire: 0.0, earth: 0.0, air: 0.0, water: 0.0)
    static let ZERO = zero
    
    init(fire: Double = 0.0, earth: Double = 0.0, air: Double = 0.0, water: Double = 0.0) {
        self.fire = fire
        self.earth = earth
        self.air = air
        self.water = water
    }
    
    subscript(element: Element) -> Double {
        get {
            switch element {
            case .fire: return fire
            case .earth: return earth
            case .air: return air
            case .water: return water
            }
        }
        set {
            switch element {
            case .fire: fire = newValue
            case .earth: earth = newValue
            case .air: air = newValue
            case .water: water = newValue
            }
        }
    }
    
    subscript(index: Int) -> Double {
        get {
            switch index {
            case 0: return fire
            case 1: return earth
            case 2: return air
            case 3: return water
            default: return 0.0
            }
        }
        set {
            switch index {
            case 0: fire = newValue
            case 1: earth = newValue
            case 2: air = newValue
            case 3: water = newValue
            default: break
            }
        }
    }
    
    func dominant() -> Element {
        let values = [fire, earth, air, water]
        let maxValue = values.max() ?? 0.0
        if fire == maxValue { return .fire }
        if earth == maxValue { return .earth }
        if air == maxValue { return .air }
        return .water
    }
    
    func normalize(to target: Double = 10.0) -> ElementVector {
        let maxVal = max(fire, earth, air, water)
        guard maxVal > 0 else { return self }
        let scale = target / maxVal
        return ElementVector(
            fire: (fire * scale).roundedTo(2),
            earth: (earth * scale).roundedTo(2),
            air: (air * scale).roundedTo(2),
            water: (water * scale).roundedTo(2)
        )
    }
    
    func dotProduct(with other: ElementVector) -> Double {
        (fire * other.fire + earth * other.earth + air * other.air + water * other.water) / 4.0
    }
    
    func roundedTo(_ decimals: Int) -> ElementVector {
        ElementVector(
            fire: fire.roundedTo(decimals),
            earth: earth.roundedTo(decimals),
            air: air.roundedTo(decimals),
            water: water.roundedTo(decimals)
        )
    }
    
    // MARK: - AdditiveArithmetic
    static func + (lhs: ElementVector, rhs: ElementVector) -> ElementVector {
        ElementVector(
            fire: lhs.fire + rhs.fire,
            earth: lhs.earth + rhs.earth,
            air: lhs.air + rhs.air,
            water: lhs.water + rhs.water
        )
    }
    
    static func - (lhs: ElementVector, rhs: ElementVector) -> ElementVector {
        ElementVector(
            fire: lhs.fire - rhs.fire,
            earth: lhs.earth - rhs.earth,
            air: lhs.air - rhs.air,
            water: lhs.water - rhs.water
        )
    }
    
    static func += (lhs: inout ElementVector, rhs: ElementVector) {
        lhs = lhs + rhs
    }
    
    static func -= (lhs: inout ElementVector, rhs: ElementVector) {
        lhs = lhs - rhs
    }
    
    static func * (lhs: ElementVector, rhs: Double) -> ElementVector {
        ElementVector(
            fire: lhs.fire * rhs,
            earth: lhs.earth * rhs,
            air: lhs.air * rhs,
            water: lhs.water * rhs
        )
    }
    
    static func / (lhs: ElementVector, rhs: Double) -> ElementVector {
        guard rhs != 0 else { return lhs }
        return ElementVector(
            fire: lhs.fire / rhs,
            earth: lhs.earth / rhs,
            air: lhs.air / rhs,
            water: lhs.water / rhs
        )
    }
}

// MARK: - Double Extension

extension Double {
    func roundedTo(_ decimals: Int) -> Double {
        let multiplier = pow(10.0, Double(decimals))
        return (self * multiplier).rounded() / multiplier
    }
}

// MARK: - Element Vector Presets

extension ElementVector {
    /// Element vectors for each sign (archetypal signatures)
    static func forSign(_ sign: Sign) -> ElementVector {
        switch sign {
        case .aries:     return ElementVector(fire: 4.0, earth: 0.0, air: 1.0, water: 0.0)
        case .taurus:    return ElementVector(fire: 0.0, earth: 4.0, air: 0.0, water: 1.0)
        case .gemini:    return ElementVector(fire: 1.0, earth: 0.0, air: 4.0, water: 0.0)
        case .cancer:    return ElementVector(fire: 0.0, earth: 1.0, air: 0.0, water: 4.0)
        case .leo:       return ElementVector(fire: 4.0, earth: 0.0, air: 1.0, water: 0.0)
        case .virgo:     return ElementVector(fire: 0.0, earth: 4.0, air: 0.0, water: 1.0)
        case .libra:     return ElementVector(fire: 1.0, earth: 0.0, air: 4.0, water: 0.0)
        case .scorpio:   return ElementVector(fire: 0.0, earth: 1.0, air: 0.0, water: 4.0)
        case .ophiuchus: return ElementVector(fire: 1.0, earth: 1.0, air: 2.0, water: 5.0)
        case .sagittarius: return ElementVector(fire: 4.0, earth: 0.0, air: 1.0, water: 0.0)
        case .capricorn: return ElementVector(fire: 0.0, earth: 4.0, air: 0.0, water: 1.0)
        case .aquarius:  return ElementVector(fire: 1.0, earth: 0.0, air: 4.0, water: 0.0)
        case .pisces:    return ElementVector(fire: 0.0, earth: 1.0, air: 0.0, water: 4.0)
        }
    }
    
    /// House elemental biases (Equal house system)
    static func forHouse(_ house: House) -> ElementVector {
        switch house {
        case .first:   return ElementVector(fire: 3.0, earth: 0.0, air: 1.0, water: 0.0)
        case .second:  return ElementVector(fire: 0.0, earth: 4.0, air: 0.0, water: 1.0)
        case .third:   return ElementVector(fire: 1.0, earth: 0.0, air: 4.0, water: 0.0)
        case .fourth:  return ElementVector(fire: 0.0, earth: 1.0, air: 0.0, water: 4.0)
        case .fifth:   return ElementVector(fire: 4.0, earth: 0.0, air: 1.0, water: 0.0)
        case .sixth:   return ElementVector(fire: 0.0, earth: 4.0, air: 0.0, water: 1.0)
        case .seventh: return ElementVector(fire: 1.0, earth: 0.0, air: 4.0, water: 0.0)
        case .eighth:  return ElementVector(fire: 0.0, earth: 1.0, air: 0.0, water: 4.0)
        case .ninth:   return ElementVector(fire: 4.0, earth: 0.0, air: 1.0, water: 0.0)
        case .tenth:   return ElementVector(fire: 0.0, earth: 4.0, air: 0.0, water: 1.0)
        case .eleventh: return ElementVector(fire: 1.0, earth: 0.0, air: 4.0, water: 0.0)
        case .twelfth:  return ElementVector(fire: 0.0, earth: 1.0, air: 0.0, water: 4.0)
        }
    }
    
    /// Phase deltas for nightly transit scoring
    static func phaseDelta(_ phase: MoonPhase) -> ElementVector {
        switch phase {
        case .newMoon:         return ElementVector(fire: 1.0, earth: 0.0, air: 0.0, water: 2.0)
        case .waxingCrescent:  return ElementVector(fire: 1.0, earth: 0.0, air: 1.0, water: 1.0)
        case .firstQuarter:    return ElementVector(fire: 2.0, earth: 0.0, air: 1.0, water: 0.0)
        case .waxingGibbous:   return ElementVector(fire: 1.0, earth: 1.0, air: 0.0, water: 1.0)
        case .fullMoon:        return ElementVector(fire: 0.0, earth: 0.0, air: 2.0, water: 2.0)
        case .waningGibbous:   return ElementVector(fire: 0.0, earth: 1.0, air: 1.0, water: 1.0)
        case .lastQuarter:     return ElementVector(fire: 0.0, earth: 2.0, air: 0.0, water: 1.0)
        case .waningCrescent:  return ElementVector(fire: 0.0, earth: 1.0, air: 0.0, water: 3.0)
        }
    }
    
    /// Transit deltas for a planet/aspect combination
    static func transitDelta(planet: Planet, aspect: Aspect) -> ElementVector? {
        let deltas: [Planet: [Aspect: ElementVector]] = [
            .moon: [
                .conjunction: ElementVector(fire: 0.0, earth: 0.0, air: 0.0, water: 3.0),
                .trine: ElementVector(fire: 0.0, earth: 0.0, air: 1.0, water: 2.0),
                .square: ElementVector(fire: 2.0, earth: 0.0, air: 0.0, water: 1.0),
                .sextile: ElementVector(fire: 0.0, earth: 0.0, air: 1.0, water: 1.5)
            ],
            .venus: [
                .conjunction: ElementVector(fire: 0.0, earth: 1.0, air: 1.0, water: 2.0),
                .trine: ElementVector(fire: 0.0, earth: 1.0, air: 0.0, water: 1.0),
                .sextile: ElementVector(fire: 0.5, earth: 0.5, air: 0.5, water: 1.0)
            ],
            .mars: [
                .conjunction: ElementVector(fire: 3.0, earth: 0.0, air: 1.0, water: 0.0),
                .square: ElementVector(fire: 3.0, earth: 0.0, air: 0.0, water: -1.0),
                .trine: ElementVector(fire: 2.0, earth: 0.0, air: 0.5, water: 0.0)
            ],
            .jupiter: [
                .conjunction: ElementVector(fire: 2.0, earth: 1.0, air: 1.0, water: 1.0),
                .trine: ElementVector(fire: 1.0, earth: 1.0, air: 1.0, water: 1.0),
                .sextile: ElementVector(fire: 1.5, earth: 0.5, air: 1.0, water: 0.5)
            ],
            .saturn: [
                .conjunction: ElementVector(fire: 0.0, earth: 3.0, air: 0.0, water: 0.0),
                .trine: ElementVector(fire: 0.0, earth: 2.0, air: 0.0, water: 0.0),
                .square: ElementVector(fire: -1.0, earth: 2.0, air: 0.0, water: 0.0)
            ],
            .uranus: [
                .conjunction: ElementVector(fire: 1.0, earth: 0.0, air: 3.0, water: 0.0),
                .trine: ElementVector(fire: 0.5, earth: 0.0, air: 2.0, water: 0.0)
            ],
            .neptune: [
                .conjunction: ElementVector(fire: 0.0, earth: 0.0, air: 1.0, water: 3.0),
                .trine: ElementVector(fire: 0.0, earth: 0.0, air: 1.0, water: 2.0),
                .sextile: ElementVector(fire: 0.0, earth: 0.0, air: 0.5, water: 1.5)
            ],
            .pluto: [
                .conjunction: ElementVector(fire: 0.0, earth: 0.0, air: 0.0, water: 4.0),
                .trine: ElementVector(fire: 0.0, earth: 0.0, air: 0.0, water: 2.5),
                .square: ElementVector(fire: 0.0, earth: 0.0, air: 0.0, water: 2.0)
            ],
            .chiron: [
                .conjunction: ElementVector(fire: 0.5, earth: 0.0, air: 1.0, water: 2.0),
                .trine: ElementVector(fire: 0.0, earth: 0.0, air: 0.5, water: 1.5)
            ],
            .lilith: [
                .conjunction: ElementVector(fire: 0.0, earth: 0.0, air: 0.0, water: 1.5),
                .square: ElementVector(fire: 0.0, earth: 0.0, air: 0.0, water: 1.0)
            ],
            .northNode: [
                .conjunction: ElementVector(fire: 0.5, earth: 0.5, air: 0.5, water: 1.0)
            ]
        ]
        
        return deltas[planet]?[aspect]
    }
}

File: AstroSleep-Android/app/src/main/java/com/astrosleep/app/core/model/ElementVector.kt
package com.astrosleep.app.core.model

import kotlinx.serialization.Serializable
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.round

/**
 * [Fire, Earth, Air, Water] scoring vector — port of iOS ElementVector.swift.
 * Mutable so nightly score accumulation can mirror iOS `+=` / subscript mutation.
 */
@Serializable
data class ElementVector(
    var fire: Double = 0.0,
    var earth: Double = 0.0,
    var air: Double = 0.0,
    var water: Double = 0.0,
) {
    operator fun get(element: Element): Double = when (element) {
        Element.FIRE -> fire
        Element.EARTH -> earth
        Element.AIR -> air
        Element.WATER -> water
    }

    operator fun set(element: Element, value: Double) {
        when (element) {
            Element.FIRE -> fire = value
            Element.EARTH -> earth = value
            Element.AIR -> air = value
            Element.WATER -> water = value
        }
    }

    operator fun get(index: Int): Double = when (index) {
        0 -> fire
        1 -> earth
        2 -> air
        3 -> water
        else -> 0.0
    }

    operator fun set(index: Int, value: Double) {
        when (index) {
            0 -> fire = value
            1 -> earth = value
            2 -> air = value
            3 -> water = value
        }
    }

    fun dominant(): Element {
        val maxValue = max(max(fire, earth), max(air, water))
        return when (maxValue) {
            fire -> Element.FIRE
            earth -> Element.EARTH
            air -> Element.AIR
            else -> Element.WATER
        }
    }

    /** Normalize so the strongest element equals [target] (iOS default 10). */
    fun normalize(target: Double = 10.0): ElementVector {
        val maxVal = max(max(fire, earth), max(air, water))
        if (maxVal <= 0.0) return copy()
        val scale = target / maxVal
        return ElementVector(
            fire = (fire * scale).roundedTo(2),
            earth = (earth * scale).roundedTo(2),
            air = (air * scale).roundedTo(2),
            water = (water * scale).roundedTo(2),
        )
    }

    /** iOS `dotProduct` — average of component products. */
    fun dotProduct(other: ElementVector): Double =
        (fire * other.fire + earth * other.earth + air * other.air + water * other.water) / 4.0

    fun roundedTo(decimals: Int): ElementVector = ElementVector(
        fire = fire.roundedTo(decimals),
        earth = earth.roundedTo(decimals),
        air = air.roundedTo(decimals),
        water = water.roundedTo(decimals),
    )

    operator fun plus(other: ElementVector): ElementVector = ElementVector(
        fire = fire + other.fire,
        earth = earth + other.earth,
        air = air + other.air,
        water = water + other.water,
    )

    operator fun minus(other: ElementVector): ElementVector = ElementVector(
        fire = fire - other.fire,
        earth = earth - other.earth,
        air = air - other.air,
        water = water - other.water,
    )

    operator fun times(scalar: Double): ElementVector = ElementVector(
        fire = fire * scalar,
        earth = earth * scalar,
        air = air * scalar,
        water = water * scalar,
    )

    operator fun div(scalar: Double): ElementVector {
        if (scalar == 0.0) return copy()
        return ElementVector(
            fire = fire / scalar,
            earth = earth / scalar,
            air = air / scalar,
            water = water / scalar,
        )
    }

    companion object {
        val ZERO = ElementVector()
        val zero get() = ZERO

        fun forSign(sign: Sign): ElementVector = when (sign) {
            Sign.ARIES -> ElementVector(4.0, 0.0, 1.0, 0.0)
            Sign.TAURUS -> ElementVector(0.0, 4.0, 0.0, 1.0)
            Sign.GEMINI -> ElementVector(1.0, 0.0, 4.0, 0.0)
            Sign.CANCER -> ElementVector(0.0, 1.0, 0.0, 4.0)
            Sign.LEO -> ElementVector(4.0, 0.0, 1.0, 0.0)
            Sign.VIRGO -> ElementVector(0.0, 4.0, 0.0, 1.0)
            Sign.LIBRA -> ElementVector(1.0, 0.0, 4.0, 0.0)
            Sign.SCORPIO -> ElementVector(0.0, 1.0, 0.0, 4.0)
            Sign.OPHIUCHUS -> ElementVector(1.0, 1.0, 2.0, 5.0)
            Sign.SAGITTARIUS -> ElementVector(4.0, 0.0, 1.0, 0.0)
            Sign.CAPRICORN -> ElementVector(0.0, 4.0, 0.0, 1.0)
            Sign.AQUARIUS -> ElementVector(1.0, 0.0, 4.0, 0.0)
            Sign.PISCES -> ElementVector(0.0, 1.0, 0.0, 4.0)
        }

        fun forHouse(house: House): ElementVector = when (house) {
            House.FIRST -> ElementVector(3.0, 0.0, 1.0, 0.0)
            House.SECOND -> ElementVector(0.0, 4.0, 0.0, 1.0)
            House.THIRD -> ElementVector(1.0, 0.0, 4.0, 0.0)
            House.FOURTH -> ElementVector(0.0, 1.0, 0.0, 4.0)
            House.FIFTH -> ElementVector(4.0, 0.0, 1.0, 0.0)
            House.SIXTH -> ElementVector(0.0, 4.0, 0.0, 1.0)
            House.SEVENTH -> ElementVector(1.0, 0.0, 4.0, 0.0)
            House.EIGHTH -> ElementVector(0.0, 1.0, 0.0, 4.0)
            House.NINTH -> ElementVector(4.0, 0.0, 1.0, 0.0)
            House.TENTH -> ElementVector(0.0, 4.0, 0.0, 1.0)
            House.ELEVENTH -> ElementVector(1.0, 0.0, 4.0, 0.0)
            House.TWELFTH -> ElementVector(0.0, 1.0, 0.0, 4.0)
        }

        fun phaseDelta(phase: MoonPhase): ElementVector = when (phase) {
            MoonPhase.NEW_MOON -> ElementVector(1.0, 0.0, 0.0, 2.0)
            MoonPhase.WAXING_CRESCENT -> ElementVector(1.0, 0.0, 1.0, 1.0)
            MoonPhase.FIRST_QUARTER -> ElementVector(2.0, 0.0, 1.0, 0.0)
            MoonPhase.WAXING_GIBBOUS -> ElementVector(1.0, 1.0, 0.0, 1.0)
            MoonPhase.FULL_MOON -> ElementVector(0.0, 0.0, 2.0, 2.0)
            MoonPhase.WANING_GIBBOUS -> ElementVector(0.0, 1.0, 1.0, 1.0)
            MoonPhase.LAST_QUARTER -> ElementVector(0.0, 2.0, 0.0, 1.0)
            MoonPhase.WANING_CRESCENT -> ElementVector(0.0, 1.0, 0.0, 3.0)
        }

        fun transitDelta(planet: Planet, aspect: Aspect): ElementVector? {
            val table: Map<Planet, Map<Aspect, ElementVector>> = mapOf(
                Planet.MOON to mapOf(
                    Aspect.CONJUNCTION to ElementVector(0.0, 0.0, 0.0, 3.0),
                    Aspect.TRINE to ElementVector(0.0, 0.0, 1.0, 2.0),
                    Aspect.SQUARE to ElementVector(2.0, 0.0, 0.0, 1.0),
                    Aspect.SEXTILE to ElementVector(0.0, 0.0, 1.0, 1.5),
                ),
                Planet.VENUS to mapOf(
                    Aspect.CONJUNCTION to ElementVector(0.0, 1.0, 1.0, 2.0),
                    Aspect.TRINE to ElementVector(0.0, 1.0, 0.0, 1.0),
                    Aspect.SEXTILE to ElementVector(0.5, 0.5, 0.5, 1.0),
                ),
                Planet.MARS to mapOf(
                    Aspect.CONJUNCTION to ElementVector(3.0, 0.0, 1.0, 0.0),
                    Aspect.SQUARE to ElementVector(3.0, 0.0, 0.0, -1.0),
                    Aspect.TRINE to ElementVector(2.0, 0.0, 0.5, 0.0),
                ),
                Planet.JUPITER to mapOf(
                    Aspect.CONJUNCTION to ElementVector(2.0, 1.0, 1.0, 1.0),
                    Aspect.TRINE to ElementVector(1.0, 1.0, 1.0, 1.0),
                    Aspect.SEXTILE to ElementVector(1.5, 0.5, 1.0, 0.5),
                ),
                Planet.SATURN to mapOf(
                    Aspect.CONJUNCTION to ElementVector(0.0, 3.0, 0.0, 0.0),
                    Aspect.TRINE to ElementVector(0.0, 2.0, 0.0, 0.0),
                    Aspect.SQUARE to ElementVector(-1.0, 2.0, 0.0, 0.0),
                ),
                Planet.URANUS to mapOf(
                    Aspect.CONJUNCTION to ElementVector(1.0, 0.0, 3.0, 0.0),
                    Aspect.TRINE to ElementVector(0.5, 0.0, 2.0, 0.0),
                ),
                Planet.NEPTUNE to mapOf(
                    Aspect.CONJUNCTION to ElementVector(0.0, 0.0, 1.0, 3.0),
                    Aspect.TRINE to ElementVector(0.0, 0.0, 1.0, 2.0),
                    Aspect.SEXTILE to ElementVector(0.0, 0.0, 0.5, 1.5),
                ),
                Planet.PLUTO to mapOf(
                    Aspect.CONJUNCTION to ElementVector(0.0, 0.0, 0.0, 4.0),
                    Aspect.TRINE to ElementVector(0.0, 0.0, 0.0, 2.5),
                    Aspect.SQUARE to ElementVector(0.0, 0.0, 0.0, 2.0),
                ),
                Planet.CHIRON to mapOf(
                    Aspect.CONJUNCTION to ElementVector(0.5, 0.0, 1.0, 2.0),
                    Aspect.TRINE to ElementVector(0.0, 0.0, 0.5, 1.5),
                ),
                Planet.LILITH to mapOf(
                    Aspect.CONJUNCTION to ElementVector(0.0, 0.0, 0.0, 1.5),
                    Aspect.SQUARE to ElementVector(0.0, 0.0, 0.0, 1.0),
                ),
                Planet.NORTH_NODE to mapOf(
                    Aspect.CONJUNCTION to ElementVector(0.5, 0.5, 0.5, 1.0),
                ),
            )
            return table[planet]?.get(aspect)
        }
    }
}

fun Double.roundedTo(decimals: Int): Double {
    val multiplier = 10.0.pow(decimals)
    return round(this * multiplier) / multiplier
}
